lighthouse-core/gather/gather-runner.js

```diff
@@ -64,26 +63,47 @@ class GatherRunner {
 
   /**
    * Loads options.url with specified options.
+   * @param {!Progress} progress
    * @param {!Driver} driver
    * @param {!Object} options
    * @return {!Promise}
    */
-  static loadPage(driver, options) {
-    return Promise.resolve()
-      // Begin tracing only if requested by config.
-      .then(_ => options.config.recordTrace && driver.beginTrace())
-      // Network is always recorded for internal use, even if not saved as artifact.
-      .then(_ => driver.beginNetworkCollect(options))
-      // Navigate.
-      .then(_ => driver.gotoURL(options.url, {
-        waitForLoad: true,
-        disableJavaScript: !!options.disableJavaScript,
-        flags: options.flags,
-      }));
+  static loadPage(progress, driver, options) {
+    return new Promise((resolve, reject) => {
+      // Set up a watch dog to reject upon progress cancel.
+      var interval = setInterval(_ => {
+        if (progress.isCanceled()) {
+          clearInterval(interval);
+          reject();
+        }
+      }, 100);
+
+      Promise.resolve()
+          // Begin tracing only if requested by config.
+          .then(_ => options.config.recordTrace && driver.beginTrace())
+          // Network is always recorded for internal use, even if not saved as artifact.
+          .then(_ => driver.beginNetworkCollect(options))
+          // Navigate.
+          .then(_ => driver.gotoURL(options.url, {
+            waitForLoad: true,
+            disableJavaScript: !!options.disableJavaScript,
+            flags: options.flags,
+          }))
+          .then(passThrough => {
+            clearInterval(interval);
+            resolve(passThrough);
+          });
+    });
   }
 
-  static setupDriver(driver, options) {
-    log.log('status', 'Initializing…');
+  /**
+   * @param {!Progress} progress
+   * @param {!Driver} driver
+   * @param {!Object} options
+   * @return {!Promise}
+   */
+  static setupDriver(progress, driver, options) {
```

Comment:
another required argument that we sometimes pass null to makes me think that we should consider using "named" arguments here via a config object instead of jamming more 

(null,null,....) values into our code which fills like a losing battle